Regression in do_not_recommend span placement in macro-generated code

This bisects to https://github.com/rust-lang/rust/pull/156676, which was designed to solve https://github.com/rust-lang/rust/issues/156475. The original issue involves code that is not macro-generated. I believe heuristics may need to be adjusted for the case of macro-generated code. It is common that macros use `diagnostic` attributes and assorted tricks to place custom error messages at some intentional location within the macro's input.

The following is minimized from https://github.com/dtolnay/cxx. It is designed to produce an error if and only if `Unpin` is not implemented for some type, in this case `Arg`.

```toml
# Cargo.toml

[package]
name = "repro"
edition = "2024"
publish = false

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1"
quote = "1"
```

```rust
// src/main.rs

pub struct Arg(std::marker::PhantomPinned);

#[repro::repro]
fn f(arg: &mut Arg);

fn main() {}
```

```rust
// src/lib.rs

use proc_macro::{TokenStream, TokenTree};

#[proc_macro_attribute]
pub fn repro(_args: TokenStream, input: TokenStream) -> TokenStream {
    // Parse input that looks like `fn f(arg: &mut Arg);`
    let mut input = input.into_iter();
    assert_eq!(input.next().unwrap().to_string(), "fn");
    assert_eq!(input.next().unwrap().to_string(), "f");
    let TokenTree::Group(group) = input.next().unwrap() else {
        unreachable!()
    };
    let mut input = group.stream().into_iter();
    assert_eq!(input.next().unwrap().to_string(), "arg");
    assert_eq!(input.next().unwrap().to_string(), ":");
    let arg: TokenStream = input.collect();
    let arg: proc_macro2::TokenStream = arg.into();

    // Emit output:
    TokenStream::from(quote::quote! {
        const _: fn() = {
            #[diagnostic::on_unimplemented(
                message = "mutable reference to C++ type requires a pin -- use Pin<&mut Arg>",
                label = "use `Pin<&mut Arg>`"
            )]
            trait ReferenceToUnpin_Arg {
                fn check_unpin() {}
            }
            #[diagnostic::do_not_recommend]
            impl<
                'a,
                T: ?::core::marker::Sized + ::core::marker::Unpin,
            > ReferenceToUnpin_Arg for &'a mut T {}
            <#arg as ReferenceToUnpin_Arg>::check_unpin
        };
    })
}
```

Before https://github.com/rust-lang/rust/pull/156676:

```console
error[E0277]: mutable reference to C++ type requires a pin -- use Pin<&mut Arg>
 --> src/main.rs:6:11
  |
6 | fn f(arg: &mut Arg);
  |           ^^^^^^^^ use `Pin<&mut Arg>`
  |
  = help: the trait `ReferenceToUnpin_Arg` is not implemented for `&mut Arg`
```

After https://github.com/rust-lang/rust/pull/156676:

```console
error[E0277]: mutable reference to C++ type requires a pin -- use Pin<&mut Arg>
 --> src/main.rs:5:1
  |
5 | #[repro::repro]
  | ^^^^^^^^^^^^^^^ use `Pin<&mut Arg>`
  |
  = help: the trait `ReferenceToUnpin_Arg` is not implemented for `&mut Arg`
  = note: this error originates in the attribute macro `repro::repro` (in Nightly builds, run with -Z macro-backtrace for more info)
```

@qaijuang @chenyukang

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
@@ -827,6 +827,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                     // Keep more precise spans that still point within the parent obligation,
                     // but do not let hidden impl details move the span outside of it.
                     if code == *root_obligation.cause.code()
+                        && root_obligation.cause.span.eq_ctxt(obligation.cause.span)
                         && !root_obligation.cause.span.contains(obligation.cause.span)
                     {
                         obligation.cause.span = root_obligation.cause.span;

--- a/tests/ui/diagnostic_namespace/do_not_recommend/auxiliary/proc_macro_repro.rs
+++ b/tests/ui/diagnostic_namespace/do_not_recommend/auxiliary/proc_macro_repro.rs
@@ -1,0 +1,37 @@
+#![feature(proc_macro_quote)]
+
+extern crate proc_macro;
+
+use proc_macro::{TokenStream, TokenTree, quote};
+
+#[proc_macro_attribute]
+pub fn repro(_args: TokenStream, input: TokenStream) -> TokenStream {
+    // Parse input that looks like `fn f(arg: &mut Arg);`
+    let mut input = input.into_iter();
+    assert_eq!(input.next().unwrap().to_string(), "fn");
+    assert_eq!(input.next().unwrap().to_string(), "f");
+    let TokenTree::Group(group) = input.next().unwrap() else { unreachable!() };
+    let mut input = group.stream().into_iter();
+    assert_eq!(input.next().unwrap().to_string(), "arg");
+    assert_eq!(input.next().unwrap().to_string(), ":");
+    let arg: TokenStream = input.collect();
+
+    // Emit output:
+    quote! {
+        const _: fn() = {
+            #[diagnostic::on_unimplemented(
+                message = "mutable reference to C++ type requires a pin -- use Pin<&mut Arg>",
+                label = "use `Pin<&mut Arg>`"
+            )]
+            trait ReferenceToUnpin_Arg {
+                fn check_unpin() {}
+            }
+            #[diagnostic::do_not_recommend]
+            impl<
+                'a,
+                T: ?::core::marker::Sized + ::core::marker::Unpin,
+            > ReferenceToUnpin_Arg for &'a mut T {}
+            <$arg as ReferenceToUnpin_Arg>::check_unpin
+        };
+    }
+}

--- a/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.current.stderr
+++ b/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.current.stderr
@@ -1,0 +1,11 @@
+error[E0277]: mutable reference to C++ type requires a pin -- use Pin<&mut Arg>
+  --> $DIR/proc-macro-span.rs:15:11
+   |
+LL | fn f(arg: &mut Arg);
+   |           ^^^^^^^^ use `Pin<&mut Arg>`
+   |
+   = help: the trait `ReferenceToUnpin_Arg` is not implemented for `&mut Arg`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.next.stderr
+++ b/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.next.stderr
@@ -1,0 +1,11 @@
+error[E0277]: mutable reference to C++ type requires a pin -- use Pin<&mut Arg>
+  --> $DIR/proc-macro-span.rs:15:11
+   |
+LL | fn f(arg: &mut Arg);
+   |           ^^^^^^^^ use `Pin<&mut Arg>`
+   |
+   = help: the trait `ReferenceToUnpin_Arg` is not implemented for `&mut Arg`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.rs
+++ b/tests/ui/diagnostic_namespace/do_not_recommend/proc-macro-span.rs
@@ -1,0 +1,16 @@
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@ compile-flags: --crate-type=lib
+//@[next] compile-flags: -Znext-solver
+//@ edition: 2024
+//@ proc-macro: proc_macro_repro.rs
+
+// Regression test for https://github.com/rust-lang/rust/issues/156759.
+
+extern crate proc_macro_repro;
+
+struct Arg(std::marker::PhantomPinned);
+
+#[proc_macro_repro::repro]
+fn f(arg: &mut Arg);
+//~^ ERROR mutable reference to C++ type requires a pin

